src/feat/signal-distort.cc

```diff
@@ -72,5 +72,87 @@ void TimeStretch(const MatrixBase<BaseFloat> &input_egs,
   perturb_egs->CopyFromMat(out_mat);
 }
 
+PerturbXvectorSignal::PerturbXvectorSignal(XvectorPerturbOptions opts): opts_(opts) {
+  if (!opts_.add_noise.empty()) {
+    // initialize the noise_list_
+    SequentialBaseFloatMatrixReader noise_seq_reader(opts_.add_noise);
+    for (; !noise_seq_reader.Done(); noise_seq_reader.Next()) {
+      std::string key = noise_seq_reader.Key();
+      noise_list_.push_back(key);
+    }
+    noise_seq_reader.Close();
+  }
+}
+// This function add the noise to the orginial signal. We should not normalize 
+// the signal level of the orginial signal. According to SNR, we rescale the noise
+// and add it. So that the perturbed signal is created. 
+void PerturbXvectorSignal::ApplyAdditiveNoise(const MatrixBase<BaseFloat> &input_eg,
```

Comment:
It is better to have separate function for applying noise than XvectorPerturbSignal class.
In the next step, you need to add extra binary as perturb-wav in featbin, which processes noise-range and apply it to input wave.
pertub-wav --noise=scp:noise.scp --noise-range="single line from range-file you generated" wav perturbed_wav
You can have struct NoiseRange with options in the range and Then it initialize  std::vector<NoiseRange> noise_ranges using range-line and define function ApplyNoise(std::string noise_scp, const std::vector<NoiseRange> noise_ranges, const VectorBase<BasFloat> &input_wav,  VectorBase<BaseFloat> *pertub_wav)
You need to check sampling frequency and ..
Then you need to add perturbed version of wav to wav.scp as perturb{perturb_id}-wavname